src/include/storage/copier/node_copier.h

```diff
@@ -1,136 +1,235 @@
+class CSVNodeCopyMorsel : public NodeCopyMorsel {
+public:
+    CSVNodeCopyMorsel(common::offset_t startOffset, std::string filePath,
+        std::shared_ptr<arrow::RecordBatch> recordBatch)
+        : NodeCopyMorsel{startOffset, BLOCK_IDX_INVALID, UINT64_MAX, std::move(filePath)},
+          recordBatch{std::move(recordBatch)} {}
+
     std::shared_ptr<arrow::RecordBatch> recordBatch;
 };
 
-template<typename T>
 class NodeCopySharedState {
-
 public:
-    NodeCopySharedState(
-        std::string filePath, HashIndexBuilder<T>* pkIndex, common::offset_t startOffset)
-        : filePath{std::move(filePath)}, pkIndex{pkIndex}, startOffset{startOffset}, blockIdx{0} {};
-
+    NodeCopySharedState(std::vector<std::string> filePaths,
+        std::unordered_map<std::string, TableCopyExecutor::FileBlockInfo> fileBlockInfos)
+        : filePaths{std::move(filePaths)}, fileIdx{0}, blockIdx{0}, nodeOffset{0},
+          fileBlockInfos{std::move(fileBlockInfos)} {};
     virtual ~NodeCopySharedState() = default;
 
-    virtual std::unique_ptr<NodeCopyMorsel> getMorsel() = 0;
+    virtual std::unique_ptr<NodeCopyMorsel> getMorsel();
 
 public:
-    std::string filePath;
-    HashIndexBuilder<T>* pkIndex;
-    common::offset_t startOffset;
+    std::vector<std::string> filePaths;
+    common::vector_idx_t fileIdx;
+    common::offset_t nodeOffset;
 
 protected:
+    // todo: do we need this?
+    std::unordered_map<std::string, TableCopyExecutor::FileBlockInfo> fileBlockInfos;
     common::block_idx_t blockIdx;
     std::mutex mtx;
 };
 
-template<typename T>
-class CSVNodeCopySharedState : public NodeCopySharedState<T> {
-
+// For CSV file, we need to read in streaming mode, so we need to keep the reader in the shared
+// state.
+class CSVNodeCopySharedState : public NodeCopySharedState {
 public:
-    CSVNodeCopySharedState(std::string filePath, HashIndexBuilder<T>* pkIndex,
-        common::offset_t startOffset,
-        std::shared_ptr<arrow::csv::StreamingReader> csvStreamingReader)
-        : NodeCopySharedState<T>{filePath, pkIndex, startOffset}, csvStreamingReader{std::move(
-                                                                      csvStreamingReader)} {};
+    CSVNodeCopySharedState(std::vector<std::string> filePaths,
+        std::unordered_map<std::string, TableCopyExecutor::FileBlockInfo> fileBlockInfos,
+        common::CSVReaderConfig* csvReaderConfig, catalog::TableSchema* tableSchema)
+        : NodeCopySharedState{std::move(filePaths), std::move(fileBlockInfos)},
+          csvReaderConfig{csvReaderConfig}, tableSchema{tableSchema} {};
+
     std::unique_ptr<NodeCopyMorsel> getMorsel() override;
 
 private:
-    std::shared_ptr<arrow::csv::StreamingReader> csvStreamingReader;
+    common::CSVReaderConfig* csvReaderConfig;
+    catalog::TableSchema* tableSchema;
+    std::shared_ptr<arrow::csv::StreamingReader> reader;
+    std::string currentFilePath;
```

Comment:
track fileIdx instead of path

Reply:
Good catch. Actually I don' think I need to track this. Removed now.